src/element/struct.rs

```diff
@@ -0,0 +1,230 @@
+        ivf.format_struct(self).map_err(|_| std::fmt::Error)?;
+        Ok(())
+    }
+}
+
+impl Struct {
+    pub fn builder() -> StructBuilder {
+        StructBuilder::new()
+    }
+
+    pub fn clone_builder(&self) -> StructBuilder {
+        StructBuilder::with_initial_fields(&self.fields.by_index)
+    }
+
+    /// Returns an iterator over the field name/value pairs in this Struct.
+    pub fn fields(&self) -> impl Iterator<Item = (&Symbol, &Element)> {
+        self.fields
+            .iter()
+            // Here we convert from &(name, value) to (&name, &value).
+            // The former makes a stronger assertion about how the data is being stored. We don't
+            // want that to be a mandatory part of the public API.
+            .map(|(name, element)| (name, element))
+    }
+
+    fn fields_eq(&self, other: &Self) -> bool {
+        // For each field name in `self`, get the list of indexes that contain a value with that name.
+        for (field_name, field_value_indexes) in &self.fields.by_name {
+            let other_value_indexes = match other.fields.get_indexes(field_name) {
+                Some(indexes) => indexes,
+                // The other struct doesn't have a field with this name so they're not equal.
+                None => return false,
+            };
+
+            if field_value_indexes.len() != other_value_indexes.len() {
+                // The other struct has fields with the same name, but a different number of them.
+                return false;
+            }
+
+            for field_value in self.fields.get_values_at_indexes(field_value_indexes) {
+                if other
+                    .fields
+                    .get_values_at_indexes(other_value_indexes)
+                    .all(|other_value| !field_value.ion_eq(other_value))
+                {
+                    // Couldn't find an equivalent field in the other struct
+                    return false;
+                }
+            }
+        }
+
+        // If all of the above conditions hold, the two structs are equal.
+        true
+    }
+
+    /// Returns the number of fields in this Struct.
+    pub fn len(&self) -> usize {
+        self.fields.by_index.len()
+    }
+
+    /// Returns `true` if this struct has zero fields.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn iter(&self) -> FieldIterator<'_> {
+        FieldIterator::new(&self.fields.by_index)
+    }
+
+    pub fn get<A: AsSymbolRef>(&self, field_name: A) -> Option<&Element> {
+        self.fields.get_last(field_name)
+    }
+
+    pub fn get_all<A: AsSymbolRef>(&self, field_name: A) -> FieldValuesIterator<'_> {
+        self.fields.get_all(field_name)
+    }
+}
+
+// Allows `for (name, value) in &my_struct {...}` syntax
+impl<'a> IntoIterator for &'a Struct {
+    type Item = (&'a Symbol, &'a Element);
+    type IntoIter = FieldIterator<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
```

Comment:
🗺️ This impl (and its unit test) is the only new code in this module. Everything else was moved here from `mod`.